python/lsst/pipe/base/graph.py

```diff
@@ -96,18 +135,67 @@ def __init__(self, iterable=None):
     def quanta(self):
         """Iterator over quanta in a graph.
 
+        Quanta are returned in unspecified order.
+
         Yields
         ------
         taskDef : `TaskDef`
             Task definition for a Quantum.
-        quantum : `Quantum`
+        quantum : `~lsst.daf.butler.Quantum`
             Single quantum.
         """
         for taskNodes in self:
             taskDef = taskNodes.taskDef
             for quantum in taskNodes.quanta:
                 yield taskDef, quantum
 
+    def orderedTaskNodes(self):
+        """Return topologically ordered task nodes.
+
+        Yields
+        ------
+        nodes : `QuantumGraphTaskNodes`
+        """
+        # Tasks in a graph are probably topologically sorted already but there
+        # is no guarantee for that. Just re-construct Pipeline and order tasks
+        # in a pipeline using existing method.
+        nodesMap = {id(item.taskDef): item for item in self}
+        pipeline = orderPipeline(Pipeline(item.taskDef for item in self))
+        for taskDef in pipeline:
+            yield nodesMap[id(taskDef)]
+
+    def orderedQuanta(self):
+        """Return topologically ordered Quanta and their pre-requisites.
+
+        This method enumerates all Quanta in a graph and iterates over
+        them in topological order so that pre-requisites for given Quantum
+        are always returned before the Quantum.
```

Comment:
I would be happy to just call this `quanta()` or `__iter__` and drop the existing `quanta()` method.

Reply:
Existing `quanta()` is also used by ctrl_mpexec, that use case does not need topologically ordered Quanta or their dependencies. I think we should support two methods, or alternatively we can require that graph always stores topologically sorted list of quanta then two methods are not needed. I'm not very happy with `__iter__`, as it implies that object is a container of the things returned by iterator, which is not entirely true in this case. We probably need more thought about what perfect interface for a graph is, and I think it should be driven by how it is used. I think that two current methods - `quanta()` and `orderedQuanta()` - are what actually makes class API, everything else at this point can probably be considered implementation detail.

Reply:
As you might have guessed my preference would probably be make the graph always topologically sorted, but I don't want to insist on that unless we can convince ourselves that it doesn't make some other operation (e.g. incremental building or merging) much harder.

If we keep both for now, perhaps we could rename `orderedQuanta()` to something that doesn't also imply (as I agree `__iter__` does) that the iterated objects are just `Quantums`?  Maybe `traverse()`?

Reply:
OK with `traverse()`.